plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs

```diff
@@ -0,0 +1,96 @@
+    }
+
+data Frame
+    = FrameApplyFun Environment (Plain Value)    -- ^ @[V _]@
+    | FrameApplyArg Environment (Plain Term)     -- ^ @[_ N]@
+    | FrameTyInstArg (Type TyName ())            -- ^ @{_ A}@
+    | FrameUnwrap                                -- ^ @(unwrap _)@
+    | FrameWrap () (TyName ()) (Type TyName ())  -- ^ @(wrap α A _)@
+
+type Context = [Frame]
+
+extendEnvironment :: Name () -> Plain Value -> Environment -> Environment -> Environment
+extendEnvironment argName arg argEnv (Environment oldEnv) =
+    Environment $ IntMap.insert (unUnique $ nameUnique argName) (arg, argEnv) oldEnv
+
+lookupName :: Name () -> Environment -> Maybe (Plain Value, Environment)
+lookupName name (Environment env) = IntMap.lookup (unUnique $ nameUnique name) env
+
+computeCek :: Environment -> Context -> Plain Term -> EvaluationResult
+computeCek env con (TyInst _ fun ty)      = computeCek env (FrameTyInstArg ty : con) fun
+computeCek env con (Apply _ fun arg)      = computeCek env (FrameApplyArg env arg : con) fun
+computeCek env con (Wrap ann tyn ty term) = computeCek env (FrameWrap ann tyn ty : con) term
+computeCek env con (Unwrap _ term)        = computeCek env (FrameUnwrap : con) term
+computeCek env con tyAbs@TyAbs{}          = returnCek env con tyAbs
+computeCek env con lamAbs@LamAbs{}        = returnCek env con lamAbs
+computeCek env con constant@Constant{}    = returnCek env con constant
+computeCek _   _   Error{}                = EvaluationFailure
+computeCek env con var@(Var _ name)       = case lookupName name env of
+    Nothing           -> throw $ MachineException OpenTermEvaluatedMachineError var
+    Just (term, env') -> returnCek env' con term
+
+returnCek :: Environment -> Context -> Plain Value -> EvaluationResult
+returnCek _   []                             res = EvaluationSuccess res
+returnCek env (FrameTyInstArg ty      : con) fun = instantiateEvaluate env con ty fun
+returnCek env (FrameApplyArg env' arg : con) fun = computeCek env' (FrameApplyFun env fun : con) arg
+returnCek env (FrameApplyFun env' fun : con) arg = applyEvaluate env' env con fun arg
+returnCek env (FrameWrap ann tyn ty   : con) val = returnCek env con $ Wrap ann tyn ty val
+returnCek env (FrameUnwrap            : con) dat = case dat of
+    Wrap _ _ _ term -> returnCek env con term
+    term            -> throw $ MachineException NonWrapUnwrappedMachineError term
+
+-- | Instantiate a term with a type and proceed.
+-- In case of 'TyAbs' just ignore the type. Otherwise check if the term is an
+-- iterated application of a 'BuiltinName' to a list of 'Value's and, if succesful,
+-- apply the term to the type via 'TyInst'.
+instantiateEvaluate
+    :: Environment -> Context -> Type TyName () -> Plain Term -> EvaluationResult
+instantiateEvaluate env con _  (TyAbs _ _ _ body) = computeCek env con body
+instantiateEvaluate env con ty fun
+    | isJust $ termAsPrimIterApp fun = returnCek env con $ TyInst () fun ty
```

Comment:
I usually do not use `ViewPatterns`, because they're kinda sneaky and I prefer to be explicit about where I compute. It's fine for me either way, but I like the way it's currently written more.

Reply:
I like it in this kind of case because the check is conceptually "part of" the match, but I don't mind if you don't like them.

Reply:
I tried and disliked what I saw.